src/env/tock/storage.rs

```diff
@@ -277,19 +277,35 @@ impl TockUpgradeStorage {
                 return Err(StorageError::CustomError);
             }
             let range = ModRange::new(storage_ptr, storage_len);
+            // First match: If we now know if A or B, write the metadata accordingly.
             match range.start() {
-                Self::METADATA_ADDRESS_A => {
-                    locations.metadata = range;
-                }
-                Self::METADATA_ADDRESS_B => {
-                    locations.running_metadata = range;
+                Self::METADATA_ADDRESS => (),
+                Self::PARTITION_ADDRESS_A => {
+                    if metadata_range.is_empty() {
+                        return Err(StorageError::NotAligned);
+                    }
+                    locations.metadata = ModRange::new(metadata_range.start(), locations.page_size);
+                    locations.running_metadata = ModRange::new(
+                        metadata_range.start() + locations.page_size,
+                        locations.page_size,
+                    );
                 }
                 Self::PARTITION_ADDRESS_B => {
-                    core::mem::swap(&mut locations.metadata, &mut locations.running_metadata);
-                    locations.partition = locations
-                        .partition
-                        .append(range)
-                        .ok_or(StorageError::NotAligned)?
+                    if metadata_range.is_empty() {
+                        return Err(StorageError::NotAligned);
+                    }
+                    locations.running_metadata =
+                        ModRange::new(metadata_range.start(), locations.page_size);
+                    locations.metadata = ModRange::new(
+                        metadata_range.start() + locations.page_size,
+                        locations.page_size,
+                    );
+                }
+                _ => (),
+            }
+            match range.start() {
+                Self::METADATA_ADDRESS => {
+                    metadata_range = range;
```

Comment:
Why not split here and do the swap in partition B as before?

Reply:
Simplified, PTAL